lib/create/attach-to-issue-from-message.js

```diff
@@ -0,0 +1,219 @@
+const { Installation } = require('../models');
+
+async function openDialog(req, res) {
+  const { slackWorkspace } = res.locals;
+  const {
+    trigger_id,
+    team,
+    channel,
+    message,
+  } = req.body;
+
+  await slackWorkspace.client.dialog.open({
+    dialog: dialogs.addComment(message, channel, team),
+    trigger_id,
+  });
+
+  return res.send();
+}
+
+async function createCommentDialogSubmission(req, res) {
+  const { slackUser, robot } = res.locals;
+  const { response_url, submission } = req.body;
+  const { issueId, url, comment } = submission;
+
+  if (!(issueId || url)) {
+    return res.send({
+      errors: [
+        {
+          name: 'issueId',
+          error: 'You need to either select an issue or pull request in the list or supply a URL',
+        },
+        {
+          name: 'url',
+          error: 'You need to either select an issue or pull request in the list or supply a URL',
+        },
+      ],
+    });
+  }
+
+  let resource;
+  if (url) {
+    resource = githubUrl(url);
+
+    if (!(resource.type === 'issue' || resource.type === 'pull')) {
+      return res.send({
+        errors: [{
+          name: 'url',
+          error: 'The URL needs to link to either an issue or a pull request',
+        }],
+      });
+    }
+
+    const permissions = {
+      issues: 'write',
+      pull_requests: 'write',
+    };
+    const { owner, repo } = resource;
+    if (!(await Installation.assertPermissions(await robot.auth(), { owner, repo, permissions }))) {
+      const appUrl = (await robot.info()).html_url;
+      return res.send({
+        errors: [{
+          name: 'url',
+          error: `Insufficient permissions. Please make sure the Slack app is installed on this repository: ${appUrl}`,
+        }],
+      });
+    }
+  }
+
+  let data;
+  try {
+    if (url) {
+      ({ data } = await slackUser.GitHubUser.client.issues.createComment({
+        ...resource,
+        type: undefined,
+        body: comment,
+      }));
+    } else {
+      const addComment = `
+        mutation comment($id: ID!, $body: String!) {
+          addComment(input: {subjectId: $id, body: $body}) {
+            commentEdge {
+              node {
+                url
+              }
+            }
+          }
+        }`;
+      data = (await slackUser.GitHubUser.client.query(
+        addComment,
+        { id: issueId, body: comment },
+      ));
+    }
+  } catch (err) {
+    if (!(err.name && err.name === 'GraphQLError')) {
+      throw err;
+    }
+    req.log.debug({ err }, 'Could not create comment');
+
+    const appUrl = (await robot.info()).html_url;
+    return res.send({
+      errors: [{
+        name: 'issueId',
+        error: `Please make sure the Slack app has been granted issue and pull request write access on this repository: ${appUrl}`,
+      }],
+    });
+  }
+
+  let issueUrl;
+  if (data.html_url) {
+    issueUrl = data.html_url;
+  } else {
+    issueUrl = data.addComment.commentEdge.node.url;
+  }
```

Comment:
Yeah, I can give that a shot. Not sure how re-usable it will be because this dialog's way of creating a comment seems pretty unique (using either a URL and REST or a node_id and GraphQL), but at least we'll be able to reduce the size of this file that way.